doc/man/man1/sacct.1
.TH sacct "1" "Slurm Commands" "April 2015" "Slurm Commands"

.SH "NAME"
sacct \- displays accounting data for all jobs and job steps in the
Slurm job accounting log or Slurm database

.SH "SYNOPSIS"
\fBsacct\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Accounting information for jobs invoked with Slurm are either logged
in the job accounting log file or saved to the Slurm database.
.PP
The \f3sacct\fP command displays job accounting data stored in the job
accounting log file or Slurm database in a variety of forms for your
analysis.  The \f3sacct\fP command displays information on jobs, job
steps, status, and exitcodes by default.  You can tailor the output
with the use of the \f3\-\-format=\fP option to specify the fields to
be shown.
.PP
For the root user, the \f3sacct\fP command displays job accounting
data for all users, although there are options to filter the output to
report only the jobs from a specified user or group.
.PP
For the non\-root user, the \f3sacct\fP command limits the display of
job accounting data to jobs that were launched with their own user
identifier (UID) by default.  Data for other users can be displayed
with the \f3\-\-allusers\fP, \f3\-\-user\fP, or \f3\-\-uid\fP options.
.TP "7"
\f3Note: \fP\c
If designated, the slurmdbd.conf option PrivateData may further
restrict the accounting data visible to users which are not
SlurmUser, root, or a user with AdminLevel=Admin. See the
slurmdbd.conf man page for additional details on restricting
access to accounting data.
.TP
\f3Note: \fP\c
If the AccountingStorageType is set to "accounting_storage/filetxt",
space characters embedded within account names, job names, and step names
will be replaced by underscores. If account names with embedded spaces are
needed, it is recommended that a database type of accounting storage be
configured.
.TP
\f3Note: \fP\c
The content's of Slurm's database are maintained in lower case. This may
result in some \f3sacct\fP output differing from that of other Slurm commands.
.TP
\f3Note: \fP\c
Much of the data reported by \f3sacct\fP has been generated by
the \f2wait3()\fP and \f2getrusage()\fP system calls. Some systems
gather and report incomplete information for these calls;
\f3sacct\fP reports values of 0 for this missing data. See your systems
\f2getrusage (3)\fP man page for information about which data are
actually available on your system.
.IP
Elapsed time fields are presented as
[days-]hours:minutes:seconds[.microseconds].  Only 'CPU' fields will
ever have microseconds.
.IP
The default input file is the file named in the
\f3AccountingStorageLoc\fP parameter in slurm.conf.

.SH "OPTIONS"

.TP "10"
\f3\-a\fP\f3,\fP \f3\-\-allusers\fP
Displays all users jobs when run by user root or if \fBPrivateData\fP is not
configured to \fBjobs\fP.
Otherwise display the current user's jobs
.IP

.TP
\f3\-A \fP\f2account_list\fP \fP\f3,\fP \f3\-\-accounts\fP\f3=\fP\f2account_list\fP
Displays jobs when a comma separated list of accounts are given as the
argument.
.IP

.TP
\f3\-b\fP\f3,\fP \f3\-\-brief\fP
Displays a brief listing, which includes the following data:
.RS
.TP "3"
\f3jobid\fP
.TP "3"
\f3status\fP
.TP "3"
\f3exitcode\fP
.RE
.IP

.TP
\f3\-c\fP\f3,\fP \f3\-\-completion\fP
Use job completion instead of job accounting.  The \f3JobCompType\fP
parameter in the slurm.conf file must be defined to a non-none option.
.IP

.TP
\f3\-\-delimiter\f3=\fP\f2characters\fP
ASCII characters used to separate the fields when specifying
the \f3\-p\fP or \f3\-P\fP options. The default delimiter
is a '|'. This options is ignored if \f3\-p\fP or \f3\-P\fP options
are not specified.

.TP
\f3\-D\fP\f3,\fP \f3\-\-duplicates\fP
If Slurm job ids are reset, some job numbers will probably appear more
than once in the accounting log file but refer to different jobs.
Such jobs can be distinguished by the "submit" time stamp in the data
records.

.IP
When data for specific jobs are requested with the \-\-jobs option,
\f3sacct\fP returns the most recent job with that number. This
behavior can be overridden by specifying \-\-duplicates, in which case
all records that match the selection criteria will be returned.

.TP
\f3\-e\fP\f3,\fP \f3\-\-helpformat\fP
.IP
Print a list of fields that can be specified with the \f3\-\-format\fP option.
.IP
.RS
.PP
.nf
.ft 3
Fields available:

AllocCPUS         AllocGRES         AllocNodes        AllocTRES
Account           AssocID           AveCPU            AveCPUFreq
AveDiskRead       AveDiskWrite      AvePages          AveRSS
AveVMSize         BlockID           Cluster           Comment
ConsumedEnergy    ConsumedEnergyRaw CPUTime           CPUTimeRAW
DerivedExitCode   Elapsed           Eligible          End
ExitCode          GID               Group             JobID
JobIDRaw          JobName           Layout            MaxDiskRead
MaxDiskReadNode   MaxDiskReadTask   MaxDiskWrite      MaxDiskWriteNode
MaxDiskWriteTask  MaxPages          MaxPagesNode      MaxPagesTask
MaxRSS            MaxRSSNode        MaxRSSTask        MaxVMSize
MaxVMSizeNode     MaxVMSizeTask     MinCPU            MinCPUNode
MinCPUTask        NCPUS             NNodes            NodeList
NTasks            PackID            PackJobID         PackStepID
Priority          Partition         QOS               QOSRAW
ReqCPUFreq        ReqCPUFreqMin     ReqCPUFreqMax     ReqCPUFreqGov
ReqCPUS           ReqGRES           ReqMem            ReqNodes
ReqTRES           Reservation       ReservationId     Reserved
ResvCPU           ResvCPURAW        Start             State
Submit            Suspended         SystemCPU         Timelimit
TotalCPU          UID               User              UserCPU
WCKey             WCKeyID

.ft 1
.fi
.RE
.IP
The section titled "Job Accounting Fields" describes these fields.

.TP
\f3\-E \fP\f2end_time\fP\fP\f3,\fP \f3\-\-endtime\fP\f3=\fP\f2end_time\fP
.IP
Select jobs in any state before the specified time.  If states are given with the \-s
option return jobs in this state before this period.

Valid time formats are...
.sp
HH:MM[:SS] [AM|PM]
.br
MMDD[YY] or MM/DD[/YY] or MM.DD[.YY]
.br
MM/DD[/YY]\-HH:MM[:SS]
.br
YYYY\-MM\-DD[THH:MM[:SS]]
.IP

.TP
\f3\-f \fP\f2file\fP\f3,\fP  \f3\-\-file\fP\f3=\fP\f2file\fP
Causes the \f3sacct\fP command to read job accounting data from the
named \f2file\fP instead of the current Slurm job accounting log
file. Only applicable when running the filetxt plugin.

.TP
\f3\-g \fP\f2gid_list\fP\f3, \-\-gid=\fP\f2gid_list\fP \f3\-\-group=\fP\f2group_list\fP
Displays the statistics only for the jobs started with the GID
or the GROUP specified by the \f2gid_list\fP or the\f2group_list\fP operand, which is a comma\-separated
list.  Space characters are not allowed.
Default is no restrictions.\&.

.TP
\f3\-h\fP\f3,\fP \f3\-\-help\fP
Displays a general help message.

.TP
\f3\-i\fP\f3,\fP \f3\-\-nnodes\fP\f3=\fP\f2N\fP
Return jobs which ran on this many nodes (N = min[\-max])

.TP
\f3\-j \fP\f2job(.step)\fP \f3,\fP  \f3\-\-jobs\fP\f3=\fP\f2job(.step)\fP
Displays information about the specified job(.step) or list of job(.step)s.
.IP
The
\f2job(.step)\fP
parameter is a comma\-separated list of jobs.
Space characters are not permitted in this list.
NOTE: A step id of 'batch' will display the information about the
batch step.  The batch step information is only available after the
batch job is complete unlike regular steps which are available when
they start.
.IP
The default is to display information on all jobs.

.TP
\f3\-k\fP\f3,\fP \f3\-\-timelimit-min\fP
Only send data about jobs with this timelimit.  If used with
timelimit_max this will be the minimum timelimit of the range.
Default is no restriction.

.TP
\f3\-K\fP\f3,\fP \f3\-\-timelimit-max\fP
Ignored by itself, but if timelimit_min is set this will be the
maximum timelimit of the range.  Default is no restriction.

.TP
\f3\-l\fP\f3,\fP \f3\-\-long\fP
Equivalent to specifying:
.IP
.na
\-\-format=jobid,jobname,partition,maxvmsize,maxvmsizenode,maxvmsizetask,
avevmsize,maxrss,maxrssnode,maxrsstask,averss,maxpages,maxpagesnode,
maxpagestask,avepages,mincpu,mincpunode,mincputask,avecpu,ntasks,
alloccpus,elapsed,state,exitcode,maxdiskread,maxdiskreadnode,maxdiskreadtask,
avediskread,maxdiskwrite,maxdiskwritenode,maxdiskwritetask,avediskwrite,
allocgres,reqgres,avecpufreq,reqcpufreqmin,reqcpufreqmax,reqcpufreqgov,
packid,packjobid,packstepid
.ad

.TP
\f3\-L\fP\f3,\fP \f3\-\-allclusters\fP
Display jobs ran on all clusters. By default, only jobs ran on the
cluster from where \f3sacct\fP is called are displayed.

.TP
\f3\-M \fP\f2cluster_list\fP\f3, \-\-clusters=\fP\f2cluster_list\fP
Displays the statistics only for the jobs started on the clusters
specified by the \f2cluster_list\fP operand, which is a
comma\-separated list of clusters.  Space characters are not allowed
in the \f2cluster_list\fP. Use \-1 for all clusters.  The default is
current cluster you are executing the \f3sacct\fP command on\&.

.TP
\f3\-n\fP\f3,\fP \f3\-\-noheader\fP
No heading will be added to the output. The default action is to
display a header.
.IP

.TP
\f3\-\-noconvert\fP
Don't convert units from their original type (e.g. 2048M won't be converted to
2G).
.IP

.TP
\f3\-N \fP\f2node_list\fP\f3, \-\-nodelist=\fP\f2node_list\fP
Display jobs that ran on any of these node(s).  \f2node_list\fP can be
a ranged string.
.IP

.TP
\f3\-\-name=\fP\f2jobname_list\fP
Display jobs that have any of these name(s).
.IP

.TP
\f3\-o\fP\f3,\fP \f3\-\-format\fP
Comma separated list of fields. (use "\-\-helpformat" for a list of
available fields).

NOTE: When using the format option for listing various fields you can put a
%NUMBER afterwards to specify how many characters should be printed.

e.g. format=name%30 will print 30 characters of field name right
justified.  A %\-30 will print 30 characters left justified.

When set, the SACCT_FORMAT environment variable will override the
default format.  For example:

SACCT_FORMAT="jobid,user,account,cluster"

.TP
\f3\-p\fP\f3,\fP \f3\-\-parsable\fP
output will be '|' delimited with a '|' at the end

.TP
\f3\-P\fP\f3,\fP \f3\-\-parsable2\fP
output will be '|' delimited without a '|' at the end

.TP
\f3\-q\fP\f3,\fP \f3\-\-qos\fP
Only send data about jobs using these qos.  Default is all.

.TP
\f3\-r\fP\f3,\fP \f3\-\-partition\fP

Comma separated list of partitions to select jobs and job steps
from. The default is all partitions.

.TP
\f3\-s \fP\f2state_list\fP \f3, \-\-state\fP\f3=\fP\f2state_list\fP
Selects jobs based on their state during the time period given.
Unless otherwise specified, the start and end time will be the
current time when the \f3\-\-state\fP option is specified and
only currently running jobs can be displayed.
A start and/or end time must be specified to view information about
jobs not currently running.
The following state designators are valid and multiple state names
may be specified using comma separators. Either the short or long form of
the state name may be used (e.g. \f3CA\fP or \f3CANCELLED\fP) and the
name is case insensitive (e.g. \f3ca\fP and \f3CA\fP both work).
.RS
.TP "20"
\fBBF  BOOT_FAIL\fR
Job terminated due to launch failure, typically due to a hardware failure
(e.g. unable to boot the node or block and the job can not be requeued).
.TP
\f3CA  CANCELLED\fP
Job was explicitly cancelled by the user or system administrator.
The job may or may not have been initiated.
.TP
\f3CD  COMPLETED\fP
Job has terminated all processes on all nodes with an exit code of zero.
.TP
\f3CF  CONFIGURING\fP
Job has been allocated resources, but are waiting for them to become ready for use
(e.g. booting).
.TP
\f3CG  COMPLETING\fP
Job is in the process of completing. Some processes on some nodes may still be active.
.TP
\f3DL  DEADLINE\fP
Job missed its deadline.
.TP
\f3F   FAILED\fP
Job terminated with non\-zero exit code or other failure condition.
.TP
\f3NF  NODE_FAIL\fP
Job terminated due to failure of one or more allocated nodes.
.TP
\f3PD  PENDING\fP
Job is awaiting resource allocation. Note for a job to be selected in this state
it must have "EligibleTime" in the requested time interval or different
from "Unknown". The "EligibleTime" is displayed by the "scontrol show job" command.
For example jobs submitted with the "\-\-hold" option will have "EligibleTime=Unknown"
as they are pending indefinitely.
.TP
\fBPR  PREEMPTED\fR
Job terminated due to preemption.
.TP
\f3R   RUNNING\fP
Job currently has an allocation.
.TP
\f3RS  RESIZING\fP
Job is about to change size.
.TP
\f3S   SUSPENDED\fP
Job has an allocation, but execution has been suspended.
.TP
\f3TO  TIMEOUT\fP
Job terminated upon reaching its time limit.
.RE
.IP
The \f2state_list\fP operand is a comma\-separated list of these state
designators.  Space characters are not allowed in the
\f2state_list\fP\c

NOTE: When specifying states and no start time is given the default
starttime is 'now'.
\&.

.TP
\f3\-S\fP\f3,\fP \f3\-\-starttime\fP
Select jobs in any state after the specified time. Default is 00:00:00 of the
current day, unless '\-s' is set then the default is 'now'.  If states
are given with the '\-s' option then only jobs in this state at this
time will be returned.

Valid time formats are...
.sp
HH:MM[:SS] [AM|PM]
.br
MMDD[YY] or MM/DD[/YY] or MM.DD[.YY]
.br
MM/DD[/YY]\-HH:MM[:SS]
.br
YYYY\-MM\-DD[THH:MM[:SS]]

.TP
\f3\-T\fP\f3,\fP \f3\-\-truncate\fP
Truncate time.  So if a job started before \-\-starttime the start time
would be truncated to \-\-starttime.  The same for end time and \-\-endtime.

.TP
\f3\-u \fP\f2uid_list\fP\f3, \-\-uid=\fP\f2uid_list\fP\f3, \-\-user=\fP\f2user_list\fP
Use this comma separated list of uids or user names to select jobs to
display.  By default, the running user's uid is used.

.TP
\f3\-\-units=[KMGTP]\fP
Display values in specified unit type. Takes precedence over \-\-noconvert
option.

.TP
\f3\-\-usage\fP
Display a command usage summary.

.TP
\f3\-v\fP\f3,\fP \f3\-\-verbose\fP
Primarily for debugging purposes, report the state of various
variables during processing.

.TP
\f3\-V\fP\f3,\fP \f3\-\-version\fP
Print version.

.TP
\f3\-W \fP\f2wckey_list\fP\f3, \-\-wckeys=\fP\f2wckey_list\fP
Displays the statistics only for the jobs started on the wckeys
specified by the \f2wckey_list\fP operand, which is a comma\-separated
list of wckey names.  Space characters are not allowed in the
\f2wckey_list\fP. Default is all wckeys\&.

.TP
\f3\-x \fP\f2associd_list\fP\f3, -\-associations=\fP\f2assoc_list\fP
Displays the statistics only for the jobs running under the
association ids specified by the \f2assoc_list\fP operand, which is a
comma\-separated list of association ids.  Space characters are not
allowed in the \f2assoc_list\fP. Default is all associations\&.

.TP
\f3\-X\fP\f3,\fP \f3\-\-allocations\fP
Only show cumulative statistics for each job, not the intermediate steps.

.SS "Job Accounting Fields"
The following describes each job accounting field:
.RS
.TP "10"
\f3ALL\fP
Print all fields listed below.

.TP
\f3AllocCPUs\fP
Count of allocated CPUs. Equivalent to \f3NCPUS\fP.

.TP
\f3AllocGRES\fP
Names and counts of generic resources allocated.

.TP
\f3AllocNodes\fP
Number of nodes allocated to the job/step.  0 if the job is pending.

.TP
\f3AllocTres\fP
Trackable resources. These are the resources allocated to the job/step
after the job started running.  For pending jobs this should be blank.
For more details see AccountingStorageTRES in slurm.conf.

.TP
\f3Account\fP
Account the job ran under.

.TP
\f3AssocID\fP
Reference to the association of user, account and cluster.

.TP
\f3AveCPU\fP
Average (system + user) CPU time of all tasks in job.

.TP
\f3AveCPUFreq\fP
Average weighted CPU frequency of all tasks in job, in kHz.

.TP
\f3AveDiskRead\fP
Average number of bytes read by all tasks in job.

.TP
\f3AveDiskWrite\fP
Average number of bytes written by all tasks in job.

.TP
\f3AvePages\fP
Average number of page faults of all tasks in job.

.TP
\f3AveRSS\fP
Average resident set size of all tasks in job.

.TP
\f3AveVMSize\fP
Average Virtual Memory size of all tasks in job.

.TP
\f3BlockID\fP
Block ID, applicable to BlueGene computers only.

.TP
\f3Cluster\fP
Cluster name.

.TP
\f3Comment\fP
The job's comment string when the AccountingStoreJobComment parameter
in the slurm.conf file is set (or defaults) to YES.  The Comment
string can be modified by invoking \f3sacctmgr modify job\fP or the
specialized \f3sjobexitmod\fP command.

.TP
\f3ConsumedEnergy\fP
Total energy consumed by all tasks in job, in joules.
Note: Only in case of exclusive job allocation this value
reflects the jobs' real energy consumption.

.TP
\f3CPUTime\fP
Formatted (Elapsed time * CPU) count used by a job or step.

.TP
\f3CPUTimeRAW\fP
Unlike above non formatted (Elapsed time * CPU) count for a job or step.
Units are cpu-seconds.

.TP
\f3DerivedExitCode\fP
The highest exit code returned by the job's job steps (srun
invocations).  Following the colon is the signal that caused the
process to terminate if it was terminated by a signal.  The
DerivedExitCode can be modified by invoking \f3sacctmgr modify job\fP
or the specialized \f3sjobexitmod\fP command.

.TP
\f3Elapsed\fP
The jobs elapsed time.
.IP
The format of this fields output is as follows:
.RS
.PD "0"
.HP
\f2[DD\-[HH:]]MM:SS\fP
.PD
.RE
.IP
as defined by the following:
.RS
.TP "10"
\f2DD\fP
days
.TP
\f2hh\fP
hours
.TP
\f2mm\fP
minutes
.TP
\f2ss\fP
seconds
.RE

.TP
\f3Eligible\fP
When the job became eligible to run.

.TP
\f3End\fP
Termination time of the job. Format output is, YYYY\-MM\-DDTHH:MM:SS, unless
changed through the SLURM_TIME_FORMAT environment variable.

.TP
\f3ExitCode\fP
The exit code returned by the job script or salloc, typically as set
by the exit() function.  Following the colon is the signal that caused
the process to terminate if it was terminated by a signal.

.TP
\f3GID\fP

RELEASE_NOTES
RELEASE NOTES FOR SLURM VERSION 17.02
23 February 2017

IMPORTANT NOTES:
THE MAXJOBID IS NOW 67,108,863. ANY PRE-EXISTING JOBS WILL CONTINUE TO RUN BUT
NEW JOB IDS WILL BE WITHIN THE NEW MAXJOBID RANGE. Adjust your configured
MaxJobID value as needed to eliminate any confusion.

If using the slurmdbd (Slurm DataBase Daemon) you must update this first.
The 17.02 slurmdbd will work with Slurm daemons of version 15.08 and above.
You will not need to update all clusters at the same time, but it is very
important to update slurmdbd first and having it running before updating
any other clusters making use of it.  No real harm will come from updating
your systems before the slurmdbd, but they will not talk to each other
until you do.  Also at least the first time running the slurmdbd you need to
make sure your my.cnf file has innodb_buffer_pool_size equal to at least 64M.
You can accomplish this by adding the line

innodb_buffer_pool_size=64M

under the [mysqld] reference in the my.cnf file and restarting the mysqld. The
buffer pool size must be smaller than the size of the MySQL tmpdir. This is
needed when converting large tables over to the new database schema.

Slurm can be upgraded from version 15.08 or 16.05 to version 17.02 without loss
of jobs or other state information. Upgrading directly from an earlier version
of Slurm will result in loss of state information.

If using SPANK plugins that use the Slurm APIs, they should be recompiled when
upgrading Slurm to a new major release.

NOTE: systemd services files are installed automatically, but not enabled.
      You will need to manually enable them on the appropriate systems:
      - Controller: systemctl enable slurmctld
      - Database: systemctl enable slurmdbd
      - Compute Nodes: systemctl enable slurmd

NOTE: If you are not using Munge, but are using the "service" scripts to
      start Slurm daemons, then you will need to remove this check from the
      etc/slurm*service scripts.

NOTE: If you are upgrading with any jobs from 14.03 or earlier
      (i.e. quick upgrade from 14.03 -> 15.08 -> 17.02) you will need
      to wait until after those jobs are gone before you upgrade to 17.02.

NOTE: If you interact with any memory values in a job_submit plugin, you will
      need to test against NO_VAL64 instead of NO_VAL, and change your printf
      format as well.

HIGHLIGHTS
==========
 -- Added infrastructure for managing workload across a federation of clusters.
    (partial functionality in version 17.02, fully operational in May 2017)
 -- In order to support federated jobs, the MaxJobID configuration parameter
    default value has been reduced from 2,147,418,112 to 67,043,328 and its
    maximum value is now 67,108,863. Upon upgrading, any pre-existing jobs that
    have a job ID above the new range will continue to run and new jobs will get
    job IDs in the new range.
 -- Added "MailDomain" configuration parameter to qualify email addresses.
 -- Automatically clean up task/cgroup cpuset and devices cgroups after steps
    are completed.
 -- Added burst buffer support for job arrays. Added new SchedulerParameters
    configuration parameter of bb_array_stage_cnt=# to indicate how many pending
    tasks of a job array should be made available for burst buffer resource
    allocation.
 -- Added new sacctmgr commands: "shutdown" (shutdown the server), "list stats"
    (get server statistics) "clear stats" (clear server statistics).
 -- The database index for jobs is now 64 bits.  If you happen to be close to
    4 billion jobs in your database you will want to update your slurmctld at
    the same time as your slurmdbd to prevent roll over of this variable as
    it is 32 bit previous versions of Slurm.
 -- All memory values (in MB) are now 64 bit. Previously, nodes with more than
    2 TB of memory would not schedule or enforce memory limits correctly.
 -- Removed AIX, BlueGene/L and BlueGene/P support.
 -- Removed sched/wiki and sched/wiki2 plugins and associated code.
 -- Added PrologFlags=Serial to disable concurrent execution of prolog/epilog
    scripts.

RPMBUILD CHANGES
================
 -- Removed separate slurm_blcr package. If Slurm is built with BLCR support,
    the files will now be part of the main Slurm packages.
 -- Replace sjstat, seff and sjobexit RPM packages with a single "contribs"
    package.
 -- Prevent installation of both init scripts and systemd service files.
    Note that systemd services are not automatically enabled at this time,
    unlike the init scripts which were automatically enabled with chkconfig.

CONFIGURATION FILE CHANGES (see man appropriate man page for details)
=====================================================================
 -- Added new "TRESWeights" configuration parameter to "NodeName" line for
    calculating how busy a node is. Currently only used for federation
    configurations.
 -- Added new "PriorityFlags" option of "INCR_ONLY", which prevents a job's
    priority from being decremented.
 -- Added "SchedulerParameters" configuration parameter of "default_gbytes",
    which treats numeric only (no suffix) value for memory and tmp disk space as
    being in units of Gigabytes. Mostly for compatability with LSF.
 -- Added "BootTime" configuration parameter to knl.conf file to optimize
    resource allocations with respect to required node reboots.
 -- Added "SchedulingParameters" option of "bf_job_part_count_reserve". Jobs
    below the specified threshold will not have resources reserved for them.
 -- Added "SchedulerParameters" option of "spec_cores_first" to select
    specialized cores from the lowest rather than highest number cores and
    sockets.
 -- Added "PrologFlags" option of "Serial" to disable concurrent launch of
    Prolog and Epilog scripts.
 -- Removed "SchedulerRootFilter" configuration parameter.
 -- Added "SyscfgTimeout" parameter to knl.conf configuration file.

COMMAND CHANGES (see man pages for details)

src/sacct/sacct.c
/*****************************************************************************\
 *  sacct.c - job accounting reports for SLURM's jobacct/log plugin
 *****************************************************************************
 *
 *  Copyright (C) 2008-2009 Lawrence Livermore National Security.
 *  Copyright (C) 2005 Hewlett-Packard Development Company, L.P.
 *
 *  This file is part of SLURM, a resource management program.
 *  For details, see <https://slurm.schedmd.com/>.
 *  Please also read the included file: DISCLAIMER.
 *
 *  SLURM is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  In addition, as a special exception, the copyright holders give permission
 *  to link the code of portions of this program with the OpenSSL library under
 *  certain conditions as described in each individual source file, and
 *  distribute linked combinations including the two. You must obey the GNU
 *  General Public License in all respects for all of the code used other than
 *  OpenSSL. If you modify file(s) with this exception, you may extend this
 *  exception to your version of the file(s), but you are not obligated to do
 *  so. If you do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source files in
 *  the program, then also delete it here.
 *
 *  SLURM is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with SLURM; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
\*****************************************************************************/

#include "sacct.h"

/*
 * Globals
 */
sacct_parameters_t params;
print_field_t fields[] = {
	{10, "Account", print_fields_str, PRINT_ACCOUNT},
	{15, "AdminComment", print_fields_str, PRINT_ADMIN_COMMENT},
	{10, "AllocCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{12, "AllocGRES", print_fields_str, PRINT_ALLOC_GRES},
	{10, "AllocNodes", print_fields_str, PRINT_ALLOC_NODES},
	{10, "AllocTRES", print_fields_str, PRINT_TRESA},
	{7,  "AssocID", print_fields_uint, PRINT_ASSOCID},
	{10, "AveCPU", print_fields_str, PRINT_AVECPU},
	{10, "AveCPUFreq", print_fields_str, PRINT_ACT_CPUFREQ},
	{14, "AveDiskRead", print_fields_str, PRINT_AVEDISKREAD},
	{14, "AveDiskWrite", print_fields_str, PRINT_AVEDISKWRITE},
	{10, "AvePages", print_fields_str, PRINT_AVEPAGES},
	{10, "AveRSS", print_fields_str, PRINT_AVERSS},
	{10, "AveVMSize", print_fields_str, PRINT_AVEVSIZE},
	{16, "BlockID", print_fields_str, PRINT_BLOCKID},
	{10, "Cluster", print_fields_str, PRINT_CLUSTER},
	{14, "Comment", print_fields_str, PRINT_COMMENT},
	{14, "ConsumedEnergy", print_fields_str, PRINT_CONSUMED_ENERGY},
	{17, "ConsumedEnergyRaw", print_fields_double,
	 PRINT_CONSUMED_ENERGY_RAW},
	{10, "CPUTime", print_fields_time_from_secs, PRINT_CPU_TIME},
	{10, "CPUTimeRAW", print_fields_uint64, PRINT_CPU_TIME_RAW},
	{15, "DerivedExitCode", print_fields_str, PRINT_DERIVED_EC},
	{10, "Elapsed", print_fields_time_from_secs, PRINT_ELAPSED},
	{10, "ElapsedRaw", print_fields_uint32, PRINT_ELAPSED_RAW},
	{19, "Eligible", print_fields_date, PRINT_ELIGIBLE},
	{19, "End", print_fields_date, PRINT_END},
	{8,  "ExitCode", print_fields_str, PRINT_EXITCODE},
	{6,  "GID", print_fields_uint, PRINT_GID},
	{9,  "Group", print_fields_str, PRINT_GROUP},
	{-12, "JobID", print_fields_str, PRINT_JOBID},
	{-12, "JobIDRaw", print_fields_str, PRINT_JOBIDRAW},
	{10, "JobName", print_fields_str, PRINT_JOBNAME},
	{9,  "Layout", print_fields_str, PRINT_LAYOUT},
	{12, "MaxDiskRead", print_fields_str, PRINT_MAXDISKREAD},
	{15, "MaxDiskReadNode", print_fields_str, PRINT_MAXDISKREADNODE},
	{15, "MaxDiskReadTask", print_fields_uint, PRINT_MAXDISKREADTASK},
	{12, "MaxDiskWrite", print_fields_str, PRINT_MAXDISKWRITE},
	{16, "MaxDiskWriteNode", print_fields_str, PRINT_MAXDISKWRITENODE},
	{16, "MaxDiskWriteTask", print_fields_uint, PRINT_MAXDISKWRITETASK},
	{8,  "MaxPages", print_fields_str, PRINT_MAXPAGES},
	{12, "MaxPagesNode", print_fields_str, PRINT_MAXPAGESNODE},
	{14, "MaxPagesTask", print_fields_uint, PRINT_MAXPAGESTASK},
	{10, "MaxRSS", print_fields_str, PRINT_MAXRSS},
	{10, "MaxRSSNode", print_fields_str, PRINT_MAXRSSNODE},
	{10, "MaxRSSTask", print_fields_uint, PRINT_MAXRSSTASK},
	{10, "MaxVMSize", print_fields_str, PRINT_MAXVSIZE},
	{14, "MaxVMSizeNode", print_fields_str, PRINT_MAXVSIZENODE},
	{14, "MaxVMSizeTask", print_fields_uint, PRINT_MAXVSIZETASK},
	{10, "MinCPU", print_fields_str, PRINT_MINCPU},
	{10, "MinCPUNode", print_fields_str, PRINT_MINCPUNODE},
	{10, "MinCPUTask", print_fields_uint, PRINT_MINCPUTASK},
	{10, "NCPUS", print_fields_uint, PRINT_ALLOC_CPUS},
	{8,  "NNodes", print_fields_str, PRINT_NNODES},
	{15, "NodeList", print_fields_str, PRINT_NODELIST},
	{8,  "NTasks", print_fields_uint, PRINT_NTASKS},
	{12, "Packid", print_fields_uint, PRINT_PACKID},
	{12, "Packjobid", print_fields_uint, PRINT_PACKJOBID},
	{12, "Packstepid", print_fields_uint, PRINT_PACKSTEPID},
	{10, "Priority", print_fields_uint, PRINT_PRIO},
	{10, "Partition", print_fields_str, PRINT_PARTITION},
	{10, "QOS", print_fields_str, PRINT_QOS},
	{6,  "QOSRAW", print_fields_uint, PRINT_QOSRAW},
	{10, "ReqCPUFreq", print_fields_str, PRINT_REQ_CPUFREQ_MAX}, /* vestigial */
	{13, "ReqCPUFreqMin", print_fields_str, PRINT_REQ_CPUFREQ_MIN},
	{13, "ReqCPUFreqMax", print_fields_str, PRINT_REQ_CPUFREQ_MAX},
	{13, "ReqCPUFreqGov", print_fields_str, PRINT_REQ_CPUFREQ_GOV},
	{8,  "ReqCPUS", print_fields_uint, PRINT_REQ_CPUS},
	{12, "ReqGRES", print_fields_str, PRINT_REQ_GRES},
	{10, "ReqMem", print_fields_str, PRINT_REQ_MEM},
	{8,  "ReqNodes", print_fields_str, PRINT_REQ_NODES},
	{10, "ReqTRES", print_fields_str, PRINT_TRESR},
	{20, "Reservation",  print_fields_str, PRINT_RESERVATION},
	{8,  "ReservationId",  print_fields_uint, PRINT_RESERVATION_ID},
	{10, "Reserved", print_fields_time_from_secs, PRINT_RESV},
	{10, "ResvCPU", print_fields_time_from_secs, PRINT_RESV_CPU},
	{10, "ResvCPURAW", print_fields_uint, PRINT_RESV_CPU},
	{19, "Start", print_fields_date, PRINT_START},
	{10, "State", print_fields_str, PRINT_STATE},
	{19, "Submit", print_fields_date, PRINT_SUBMIT},
	{10, "Suspended", print_fields_time_from_secs, PRINT_SUSPENDED},
	{10, "SystemCPU", print_fields_str, PRINT_SYSTEMCPU},
	{10, "Timelimit", print_fields_str, PRINT_TIMELIMIT},
	{10, "TotalCPU", print_fields_str, PRINT_TOTALCPU},
	{6,  "UID", print_fields_uint, PRINT_UID},
	{9,  "User", print_fields_str, PRINT_USER},
	{10, "UserCPU", print_fields_str, PRINT_USERCPU},
	{10, "WCKey", print_fields_str, PRINT_WCKEY},
	{10, "WCKeyID", print_fields_uint, PRINT_WCKEYID},
	{0,  NULL, NULL, 0}};

List jobs = NULL;

int main(int argc, char **argv)
{
	enum {
		SACCT_LIST,
		SACCT_HELP,
		SACCT_USAGE
	} op;
	int rc = 0;

	slurm_conf_init(NULL);
	sacct_init();
	parse_command_line(argc, argv);

	/* What are we doing? Requests for help take highest priority,
	 * but then check for illogical switch combinations.
	 */

	if (params.opt_help)
		op = SACCT_HELP;
	else
		op = SACCT_LIST;


	switch (op) {
	case SACCT_LIST:
		print_fields_header(print_fields_list);
		if (get_data() == SLURM_ERROR)
			exit(errno);
		if (params.opt_completion)
			do_list_completion();
		else
			do_list();
		break;
	case SACCT_HELP:
		do_help();
		break;
	default:
		fprintf(stderr, "sacct bug: should never get here\n");
		sacct_fini();
		exit(2);
	}

	sacct_fini();
	return (rc);
}
